feat: Use "Log" as default event name

psandana opened the pull request:
Fixes #
Design discussion issue (if applicable) #

## Changes

Please provide a brief description of the changes here.

## Merge requirement checklist

* [ ] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-rust-contrib/blob/main/CONTRIBUTING.md) guidelines followed
* [ ] Unit tests added/updated (if applicable)
* [ ] Appropriate `CHANGELOG.md` files updated for non-trivial, user-facing changes
* [ ] Changes in public API reviewed (if applicable)

martintmk commented on opentelemetry-etw-logs/src/logs/exporter_options.rs:
```diff
@@ -0,0 +1,262 @@
+
+impl ExporterOptions {
+    /// TODO: Add documentation
+    pub fn builder(provider_name: Cow<'static, str>) -> ExporterOptionsBuilder {
```
```suggestion
    pub fn builder(provider_name: impl Into<Cow<'static, str>>) -> ExporterOptionsBuilder {
```

^^ allows passing strings directly

psandana commented on opentelemetry-etw-logs/src/logs/exporter_options.rs:
```diff
@@ -0,0 +1,265 @@
+    }
+
+    /// TODO: Add documentation
+    pub fn provider_name(&self) -> &str {
```
```suggestion
    pub(crate) fn provider_name(&self) -> &str {
```

psandana commented on opentelemetry-etw-logs/src/logs/exporter_options.rs:
```diff
@@ -0,0 +1,265 @@
+    }
+
+    /// TODO: Add documentation
+    pub fn default_event_name(&self) -> &str {
```
```suggestion
    pub(crate) fn default_event_name(&self) -> &str {
```

psandana commented on opentelemetry-etw-logs/src/logs/exporter_options.rs:
```diff
@@ -0,0 +1,265 @@
+    }
+
+    /// TODO: Add documentation
+    pub fn get_etw_event_name<'a>(
```
```suggestion
    pub(crate) fn get_etw_event_name<'a>(
```

psandana commented on opentelemetry-etw-logs/src/logs/exporter_options.rs:
```diff
@@ -0,0 +1,143 @@
+    ) -> &'a str {
+        // Using target for now. This is the default behavior.
+        // Future versions of this library may add mechanisms to chose which attribute to use for the mapping key
+        if let Some(target) = log_record.target() {
```
I propose a default behavior of using `target` if exists, and default to `Log` or any other default set by user. `log` and `tracing` crates uses `target` with defaults to contains the `crate` (and possibly the module) originating a log record. This can be used to write to the desired ETW Event Name.

While hardcoding `Logs` can work for basic scenarios, most users of ETW will expect to configure their own Event Names.

psandana replied:
After discussing this further, we are providing:
- `Log` as default value for ETW event name
- the `Log` default value can be overridden by calling: `with_default_event_name()`
- if user wants to use the `SdkLogRecord` name, it can select: `use_etw_event_name_from_name()`
- if user wants to use the `SdkLogRecord` target, it can select: `use_etw_event_name_from_target()`

cijothomas commented on opentelemetry-etw-logs/src/logs/reentrant_logprocessor.rs:
```diff
@@ -6,40 +6,29 @@ use opentelemetry_sdk::logs::{LogBatch, LogExporter, SdkLogRecord};
     }
 }
 
+/// Creates an opaque LogProcessor that can be used with the OpenTelemetry SDK.
```
given this returns a *new* instance of LogRecordProcessor, *not* the one actually used if user did `with_etw_exporter` option to add etw_exporter, this can make it easy for users to make mistakes!
(it won't be obvious that the processor returned in this method is not the same as what is used by the trait method!)

cijothomas replied:
> It should definitely be clarified better in the rustdoc, but I'd hope that that's enough. Or do you have something else in mind as for how to address this?

Doc is the minimum we should do. Or better is to avoid 2 APIs, so there is no scope of misconfiguring things.

cijothomas commented on opentelemetry-etw-logs/src/logs/exporter_options.rs:
```diff
@@ -0,0 +1,356 @@
+        }
+    }
+
+    /// Sets the default event name to use by `ExporterOptions::get_etw_event_name` if:
```
`ExporterOptions::get_etw_event_name` is an internal method right? Lets not mention about it public docs.

cijothomas commented on opentelemetry-etw-logs/src/logs/exporter_options.rs:
```diff
@@ -0,0 +1,356 @@
+        self
+    }
+
+    /// Sets the provider name to be the default value ("Log"). The default value may be overridden by `with_default_event_name()`.
```
likely a typo - you meant etw_event name, not the provider name right?

cijothomas commented on opentelemetry-etw-logs/src/logs/exporter_options.rs:
```diff
@@ -0,0 +1,356 @@
+    }
+
+    /// Validates the options given by consuming itself and returning the `ExporterOptions` or an error.
+    pub fn build(self) -> Result<ExporterOptions, String> {
```
Is it better to use a dedicated Error enum, instead of String, which the user has to parse out to understand what went wrong? as of today, only provider name can cause failure, but in the future there could be other build errors we want to surface back to users?

OTLP Example:
https://github.com/open-telemetry/opentelemetry-rust/blob/main/opentelemetry-otlp/src/exporter/mod.rs#L107

psandana replied:
Done

cijothomas commented on opentelemetry-etw-logs/src/logs/exporter_options.rs:
```diff
@@ -0,0 +1,356 @@
+    }
+
+    /// Sets the provider name to be the default value ("Log"). The default value may be overridden by `with_default_event_name()`.
+    pub fn use_etw_event_name_from_default(mut self) -> Self {
```
isn't this the default anyway, so does the user has to set default as default explicitly?

psandana replied:
Yes, it is the default already. No, use does not have to do it. It is left to document better the 3 possibilities. If we remove it, users may have the wrong assumption they need to choose from target or name.

I will improve docs around this

cijothomas commented on opentelemetry-etw-logs/Cargo.toml:
```diff
@@ -48,5 +50,9 @@ default = ["internal-logs"]
 name = "basic"
 path = "examples/basic.rs"
 
+[[example]]
```
not so sure if its worth adding another example, just for exporter_options, which is anyway basically covered in the basic.rs too?

psandana replied:
basic is just the default. I will add more options to make it worth to have it.

psandana replied:
Removed

cijothomas commented on opentelemetry-etw-logs/src/logs/with_etw_exporter.rs:
```diff
@@ -60,4 +58,28 @@ mod tests {
+
+    #[test]
+    fn log_with_etw_exporter_trait() {
+        use log::{error, Level};
```
why test `log` here specifically? Exporters should have no connection to the logging library anyway right?

psandana replied:
I agree, and for that we should use the `LogRecord` interface directly. But as we decided to not do so, and use `tracing`, I also feel we should explicitly test the integration of `log` crate for parity.

cijothomas replied:
please make those changes in separate PR to keep the scope in control. This is already touching too many aspects in one PR.

cijothomas commented on opentelemetry-etw-logs/src/logs/exporter_options.rs:
```diff
@@ -0,0 +1,356 @@
+
+/// Options used by the Exporter.
+#[derive(Debug)]
+pub struct ExporterOptions {
```
@lalitb @utpilla 
Would love to get more feedback on this approach. 
One downside with this approach is that, users will most likely need to write LogRecordProcessor to customize Name,Target as their defaults may not work well. (i.e the default name, target provided by `tracing` is not likely useful)

Other option is to expose a callback for end-users, in the exporter.
`get_etw_event_name(&log_record) -> &str`
This has the upside that users can return the name without having to replace the original name (and target), so PartB.Log.name will always contain the original EventName, and users can influence PartA-name (and Table names in backend). This does have a risk of letting arbitrary user code run in the hot path. Given the alternative is to write LogRecordProcessor, which is also arbitrary user-code, we could reconsider (@psandana yes, I known I was not very supportive of providing arbitrary callbacks, but considering the full picture, I can change my position) this approach.

cijothomas commented on opentelemetry-etw-logs/src/logs/processor.rs:
```diff
@@ -0,0 +1,295 @@
+    /// Creates a new instance of `ProcessorBuilder` with the given provider name.
+    ///
+    /// By default, all events will be exported to the "Log" ETW event.
+    pub fn new(provider_name: impl Into<Cow<'static, str>>) -> Self {
```
can you check if this can be consistent with user_events exporter, which accepts `&str` for provider_name.

cijothomas commented on opentelemetry-etw-logs/src/logs/processor.rs:
```diff
@@ -0,0 +1,295 @@
+}
+
+impl std::error::Error for ProcessorBuildError {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
```
is this required to be implemented?

cijothomas commented on opentelemetry-etw-logs/src/logs/processor.rs:
```diff
@@ -0,0 +1,295 @@
+    /// Creates a new instance of `ProcessorBuilder` with the given provider name.
+    ///
+    /// By default, all events will be exported to the "Log" ETW event.
+    pub fn new(provider_name: impl Into<Cow<'static, str>>) -> Self {
```
for consistency with user_events, lets not keep this public. Only way for users to get to this is via Processor::builder(..).

psandana replied:
Done

cijothomas commented on opentelemetry-etw-logs/src/logs/processor.rs:
```diff
@@ -0,0 +1,295 @@
+
+/// Errors that can occur while building a `Processor` instance.
+#[derive(Debug)]
+pub struct ProcessorBuildError {
```
I think we should stick with existing patterns in the Otel repo. 
We can either do an Enum with all possible errors listed (and mark non_exhaustive).
Or
return `Box<dyn Error>`

user-events exporter followed the pattern of returning `Box<dyn Error>`

cijothomas commented on opentelemetry-etw-logs/src/logs/processor.rs:
```diff
@@ -0,0 +1,295 @@
+    #[error("Provider name must contain only ASCII alphanumeric characters, '_' or '-'.")]
+    ProviderNameInvalid,
+    #[error("Event name cannot be empty.")]
+    EventNameEmpty,
```
this looks incorrect....Since event_name is not an input to the builder, builder cannot fail to build due to empty event name...

Also, this pub (crate) enum vs the pub struct is making bit hard to follow the overall flow... Unless there is a very strong reason, lets keep it simpler

cijothomas commented on opentelemetry-etw-logs/examples/basic.rs:
```diff
@@ -39,14 +39,20 @@
 fn init_logger() -> SdkLoggerProvider {
+    let processor = Processor::builder("provider-name")
+        // Use a constant custom name for all events instead of "Log"
+        .etw_event_name_from_callback(|_| "CustomEvent")
```
nit: basic/simple example need not show customization.  suggest to use the simplest.

psandana replied:
removed

cijothomas commented on opentelemetry-etw-logs/src/logs/exporter/options.rs:
```diff
@@ -0,0 +1,256 @@
+}
+
+#[derive(Debug)]
+pub(crate) struct OptionsBuilder {
```
its easier to review, if builder is handled completed inside the ProcessorBuilder itself.

cijothomas commented on Cargo.toml:
```diff
@@ -29,6 +29,7 @@ debug = 1
 
 [workspace.dependencies]
 opentelemetry = "0.29"
+opentelemetry-appender-log = "0.29"
```
please make these changes in separate PRs. Big PRs are extremely time consuming to review. Address one issue/concern at a time.

psandana replied:
This was dangling dep. Removing

cijothomas reviewed (changes requested):
Thanks! The direction looks good, here are few things we need to reconsider 
1. Please remove all changes unrelated to changing from "with_etw" to the processor builder pattern. Large PRs gets harder to review, especially when new commits are pushed.
2. Unsure if we should use a totally different approach of returning Errors here - I suggest to stick with returning Box<dyn Err> for now.
3. user-events did recent refactor to support the new processor builder, see if we can make it consistent with that. I think we should simplify code bit more to make it easier to follow, left some comments about this in couple of places.

psandana commented on opentelemetry-etw-logs/src/logs/processor.rs:
```diff
@@ -0,0 +1,295 @@
+        match self.options_builder.build() {
+            Ok(options) => Ok(Processor::new(options)),
+            Err(error) => {
+                otel_warn!(name: "ETW.Processor.CreationFailed", reason = &error.to_string());
```
removed

psandana commented on opentelemetry-etw-logs/src/logs/exporter/mod.rs:
```diff
@@ -233,6 +240,21 @@ mod tests {
+        let exporter = common::test_utils::new_etw_exporter();
+        let result = exporter.export(batch);
+
+        assert!(result.await.is_ok());
```
Good call. Fixed

psandana commented on opentelemetry-etw-logs/CHANGELOG.md:
```diff
@@ -2,8 +2,18 @@
-- Added validation to provider name
+- Added validation to provider name.
 - Added optional feature `serde_json` to serialize List and Maps.
+- Added `etw_log_processor()` method that returns an `impl LogProcessor`.
```
Done

psandana commented:
> Thanks! The direction looks good, here are few things we need to reconsider
> 
> 1. Please remove all changes unrelated to changing from "with_etw" to the processor builder pattern. Large PRs gets harder to review, especially when new commits are pushed.
> 2. Unsure if we should use a totally different approach of returning Errors here - I suggest to stick with returning Box for now.
> 3. user-events did recent refactor to support the new processor builder, see if we can make it consistent with that. I think we should simplify code bit more to make it easier to follow, left some comments about this in couple of places.

Hi @cijothomas . I reworked the PR using the feedback provided. Errors are Boxed and simplified things. Please review.

cijothomas commented on opentelemetry-etw-logs/Cargo.toml:
```diff
@@ -11,15 +11,19 @@ keywords = ["opentelemetry", "log", "trace", "etw"]
+tracing = { version = "0.1", optional = true }
 
 [dev-dependencies]
+log = { version = "0.4.21", features = ["kv_serde"] }
```
is this required?

psandana replied:
Not required anymore, along with microbench. Fixed

cijothomas commented on opentelemetry-etw-logs/Cargo.toml:
```diff
@@ -11,15 +11,19 @@ keywords = ["opentelemetry", "log", "trace", "etw"]
 
 [dependencies]
-tracelogging_dynamic = "1.2.4"
+chrono = { version = "0.4.40", default-features = false, features = [
```
its hard to tell if this file actually has changes or not..looks like just re-arranging ? please remove unrelated changes from this PR.

psandana replied:
I won't trim this PR further. Please, accept this PR as is. Subsequent ones should be small, incremental changes.

cijothomas replied:
Sorry I don't follow. Does this file actually have any changes or just re-arranging the order?
Its putting unnecessary burden on reviewers to closely check and confirm if this is actually changed or not.

utpilla replied:
> Its putting unnecessary burden on reviewers to closely check and confirm if this is actually changed or not.

I’m with @cijothomas on this. Simplifying the PR for review not only helps reviewers but also benefits the author by increasing the chances of a quicker merge. In this case, the only intended change appears to be the removal of `microbench` from the dev dependencies. Ideally, that should have been the sole modification, without reordering the other dependencies.

> The project does not have a toml formatting, causing this kind of inconsistencies.

It’s usually a good idea to stick to the existing style, formatting, or conventions used in the project when sending in PRs. If there doesn’t seem to be a clear standard yet, it might be better to open a separate PR to introduce one, rather than mixing style changes into an unrelated change.

cijothomas commented on opentelemetry-etw-logs/src/logs/processor.rs:
```diff
@@ -0,0 +1,302 @@
+    ///
+    /// The resulting name must be a valid CommonSchema 4.0 TraceLoggingDynamic event name. Otherwise,
+    /// the default "Log" ETW event name will be used.
+    pub fn etw_event_name_from_callback(
```
Please see the discussion here : https://github.com/open-telemetry/opentelemetry-rust-contrib/pull/275#discussion_r2103713620

We need to decide how best to handle this.

cijothomas commented on opentelemetry-etw-logs/src/logs/processor.rs:
```diff
@@ -0,0 +1,302 @@
+        self
+    }
+
+    /// Validates the options given by consuming itself and returning the `Processor` or an error.
```
nit: The comment should be something like "Builds the processor with given options, returning Err when build fails (eg: due to invalid options)".

cijothomas commented on opentelemetry-etw-logs/src/logs/processor.rs:
```diff
@@ -0,0 +1,302 @@
+    pub fn build(self) -> Result<Processor, Box<dyn Error>> {
+        self.validate()?;
+
+        Ok(Processor::new(self.options_builder))
```
looks a bit odd to see that a builder is still passed to Processor...is the processor expected to call build() on the options_builder()?

psandana replied:
`options_builder` should be just `options`. The change was missing from last refactor. Fixed.

cijothomas commented on opentelemetry-etw-logs/src/logs/exporter/options.rs:
```diff
@@ -0,0 +1,169 @@
+            .etw_event_name_from_callback(|log_record| log_record.event_name().unwrap_or_default());
+
+        let result = options.get_etw_event_name(&log_record);
+        assert_eq!(result, "Log");
```
interesting that "Log" is still the name, though the callback is always returning something... Is this trying to test that `invalid_event_name_from_callback_will_use_default` ?

cijothomas replied:
Got it. It's better to have a targeted test for the same. `invalid_event_name_from_callback_will_use_default` would be my suggestion.

Not a blocker.

cijothomas commented on opentelemetry-etw-logs/src/logs/exporter/options.rs:
```diff
@@ -0,0 +1,169 @@
+        let result = options.get_etw_event_name(&log_record);
+        assert_eq!(result, "Log");
+
+        log_record.set_target("target-name");
```
what happens if we pass empty string to target?

psandana replied:
Depends on the callback given. Assuming the callback is returning the target, the current validation will discard it because it is empty and return the default "Log".

psandana commented on opentelemetry-etw-logs/src/logs/processor.rs:
```diff
@@ -0,0 +1,302 @@
+    ///
+    /// - non-empty,
+    /// - less than 234 characters long, and
+    /// - contain only ASCII alphanumeric characters, underscores (`_`), or hyphens (`-`).
```
If we follow the stricter user-events validation, then we will not be able to match all the possible valid providers in ETW. There are system ETW providers that uses hyphens already (unofficial list: https://gist.github.com/guitarrapc/35a94b908bad677a7310)

psandana commented on opentelemetry-etw-logs/src/logs/processor.rs:
```diff
@@ -0,0 +1,302 @@
+        use tracing_subscriber::prelude::*;
+
+        let processor = Processor::builder("provider-name")
+            .etw_event_name_from_callback(|_| "CustomEvent")
```
It is an integration test at this moment. Renamed to `tracing_integration_test`. We can move it to the `tests` folder (we should create it) if you prefer, either on this PR or a new one.

cijothomas replied:
My comment was more about - "There are not asserts in the test, so what is the purpose of the test"?

psandana commented on opentelemetry-etw-logs/src/logs/exporter/part_c.rs:
```diff
@@ -1,9 +1,9 @@
+pub(crate) const EVENT_ID: &str = "event_id";
 
-pub fn populate_part_c(
+pub(crate) fn populate_part_c(
```
This is just more intentional and resilient to refactorings.

utpilla commented on opentelemetry-etw-logs/README.md:
````diff
@@ -28,7 +28,51 @@ captured by agents running locally and listening for specific ETW events.
+    $EtwSessionGuid = (new-object System.Diagnostics.Tracing.EventSource("provider-name")).Guid.ToString()`
+    ```
+
+1. Start Logman session:
````
Correct the numbering used for these steps. All the steps are numbered as "1".

psandana replied:
Yes. This is correct. MD solves them at rendering time. https://www.markdownguide.org/basic-syntax/#ordered-lists

utpilla replied:
Ah okay
